constraints.pro

```diff
@@ -244,6 +244,26 @@ gen_enforced_field(WorkspaceCwd, 'types', './dist/index.d.ts') :-
 gen_enforced_field(WorkspaceCwd, 'types', null) :-
   workspace_field(WorkspaceCwd, 'private', true).
 
+% The exports for all published packages must be the same.
+gen_enforced_field(WorkspaceCwd, 'exports["."].import', './dist/index.mjs') :-
+  \+ workspace_field(WorkspaceCwd, 'private', true).
+gen_enforced_field(WorkspaceCwd, 'exports["."].require', './dist/index.js') :-
+  \+ workspace_field(WorkspaceCwd, 'private', true).
+gen_enforced_field(WorkspaceCwd, 'exports["."].types', './dist/types/index.d.ts') :-
+  \+ workspace_field(WorkspaceCwd, 'private', true).
+gen_enforced_field(WorkspaceCwd, 'exports["./package.json"]', './package.json') :-
+  \+ workspace_field(WorkspaceCwd, 'private', true).
+% Non-published packages must not specify exports.
+gen_enforced_field(WorkspaceCwd, 'exports', null) :-
+  workspace_field(WorkspaceCwd, 'private', true).
+
+% Published packages must not have side effects.
+gen_enforced_field(WorkspaceCwd, 'sideEffects', false) :-
+  \+ workspace_field(WorkspaceCwd, 'private', true).
```

Comment:
I haven't checked every single package for side effects, but it looks like there aren't any. Setting `sideEffects: false` improves tree shaking in tools like Webpack.